crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+        {
+            return Some(true);
+        }
+
+        if qargs1.len() > max_num_qubits as usize || qargs2.len() > max_num_qubits as usize {
+            return Some(false);
+        }
+
+        if SUPPORTED_OP.contains(&op1.op().name()) && SUPPORTED_OP.contains(&op2.op().name()) {
+            return None;
+        }
+
+        if self.is_commutation_skipped(op1) || self.is_commutation_skipped(op2) {
+            return Some(false);
+        }
+
+        None
+    }
+
+    fn get_op(py: Python, gate: &PyObject) -> Option<Array2<Complex64>> {
+        Some(
+            QI_OPERATOR
+                .get_bound(py)
+                .call1((gate,))
+                .ok()?
+                .getattr(intern!(py, "data"))
+                .ok()?
+                .extract::<PyReadonlyArray2<Complex64>>()
+                .ok()?
+                .as_array()
+                .to_owned(),
+        )
+    }
+
+    fn is_commutation_skipped(&self, instr: &CircuitInstruction) -> bool {
+        let op = instr.op();
+        op.directive() || SKIPPED_NAMES.contains(&op.name()) || instr.is_parameterized()
+    }
+
+    fn get_relative_placement(
+        first_qargs: &[usize],
+        second_qargs: &[usize],
+    ) -> SmallVec<[Option<Qubit>; 2]> {
+        let qubits_g2: HashMap<_, _> = second_qargs
+            .iter()
+            .enumerate()
+            .map(|(i_g1, q_g1)| (q_g1, Qubit(i_g1 as u32)))
+            .collect();
+
+        first_qargs
+            .iter()
+            .map(|q_g0| qubits_g2.get(q_g0).copied())
+            .collect()
+    }
+
+    fn clear_cache(&mut self) {
+        self.cache.clear();
+        self.current_cache_entries = 0;
+        self._cache_miss = 0;
+        self._cache_hit = 0;
+        self._cache_hit = 0;
+    }
+}
+
+#[derive(Clone)]
+#[pyclass]
+pub struct CommutationLibrary {
+    pub library: HashMap<(String, String), CommutationLibraryEntry>,
+}
+
+impl CommutationLibrary {
+    fn check_commutation_entries(
+        &self,
+        first_op: &OperationRef,
+        first_qargs: &[usize],
+        second_op: &OperationRef,
+        second_qargs: &[usize],
+    ) -> Option<bool> {
+        match self
+            .library
+            .get(&(first_op.name().to_string(), second_op.name().to_string()))
+        {
+            Some(CommutationLibraryEntry::Commutes(b)) => Some(*b),
+            Some(CommutationLibraryEntry::QubitMapping(qm)) => qm
+                .get(&CommutationChecker::get_relative_placement(
+                    first_qargs,
+                    second_qargs,
+                ))
+                .copied(),
+            _ => None,
+        }
+    }
+}
+
+impl ToPyObject for CommutationLibrary {
+    fn to_object(&self, py: Python) -> PyObject {
+        self.library.to_object(py)
+    }
+}
+
+#[pymethods]
+impl CommutationLibrary {
+    #[new]
+    fn new(py: Python, py_any: Option<Py<PyAny>>) -> Self {
```

Comment:
Change the signature to `Option<Bound<PyAny>>` to receive a bound object, avoid requiring an extra `py: Python` and avoid having to call `.bind(py)` later on.

```suggestion
    fn new(py_any: Option<Bound<PyAny>>) -> Self {
```